### PR TITLE
Add SmartDataCenter/Joyent Public Cloud support.

### PR DESCRIPTION
**Pull request by bbhoss**
Initial support:
Create single instance
Create multiple instances by count
Maintain group of instances by count_tag/exact_count parameters
Terminating single instance
Terminating instances by tag(s)

### CODE REVIEW

**Comment by baetheus** [comment]
In the SDC webui one can delete a zone/vm. In SmartOS via vmadm, one can delete or destroy a zone/vm. You might consider using one of the existing terminologies instead of using terminate to reference delete/destroy.

File: cloud/smartdatacenter/sdc.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+'''
+
+import os
+import sys
+import time
+try:
+    import smartdc
+    from smartdc import DataCenter
+except ImportError:
+    print "failed=True msg='smartdc required for this module'"
```
Comment (sivel) [inline cloud/smartdatacenter/sdc.py]: This form of 'Baby JSON' was removed from ansible 1.8.  The recommended pattern is something like:

``` python
try:
    import smartdc
    HAS_SMARTDC = True
except ImportError:
    HAS_SMARTDC = False

# Other code

def main():
    module = AnsibleModule(...)
    if not HAS_SMARTDC:
        module.fail_json(msg='smartdc python module is required')
```

File: cloud/smartdatacenter/sdc.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    sys.exit(1)
+import json
```
Comment (sivel) [inline cloud/smartdatacenter/sdc.py]: There is no need to import `json` here.  You will automatically get `json` or `simplejson` from the `ansible.module_utils.basic` import.

File: cloud/smartdatacenter/sdc.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+    wait_timeout = int(module.params.get('wait_timeout'))
+    wait_timeout_time = time.time() + wait_timeout
+    while not all( sm.status() == status for sm in machines ) and wait_timeout_time > time.time():
+        time.sleep(2)
+    # Check again that all are in expected state. If not, we timed out
+    if not all( sm.status() == status for sm in machines ):
+        module.fail_json(msg="Timed out creating machines")
+
+def list_existing_machines(sdc, tags):
+    assert tags, "Must have tags to list existing machines for termination/counting"
```
Comment (sivel) [inline cloud/smartdatacenter/sdc.py]: You probably don't want this to actually raise an AssertionError exception.  You should evaluate this in a way that allows you to run `module.fail_json` instead.

**Comment by sivel** [comment]
Just as a note, it may be beneficial to use a topic branch in your repo other than `devel` when submitting pull requests.  If you make any changes to your devel branch unrelated to these changes they will show up in this pull request.

**Comment by bbhoss** [comment]
Thanks for the comments. I will get this fixed up soon. Should I remake the pull request so there is only 1 commit, or just push my commit where I make the changes up as usual? I will make a topic branch next time.

@slashtact I'm not sure I understand you. Are you just saying that I should change the terminate language to delete?

**Comment by bbhoss** [comment]
ok, I've updated my code.

**Comment by L2G** [comment]
@bbhoss Thanks for your work on this. I'm just an Ansible user, not a committer or maintainer or anything, but if you do any more SmartOS/SmartDataCenter features for Ansible, I'll be interested in checking them out.

File: cloud/smartdatacenter/sdc.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
+#!/usr/bin/env python
```
Comment (bcoca) [inline cloud/smartdatacenter/sdc.py:1]: this should be /usr/bin/python

File: cloud/smartdatacenter/sdc.py [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+DOCUMENTATION = '''
+---
+module: sdc
+short_description: create or delete a virtualmachine or smartmachine in SDC
+description:
+     - Creates or deletes SDC instances. When created optionally waits for it to be 'running'. This module has a dependency on smartdc >= 0.2.0 which is available through pip
+options:
+  machine_id:
+    description:
+     - id of machine. Used only for deletion
```
Comment (bcoca) [inline cloud/smartdatacenter/sdc.py:26]: is this requied?

File: cloud/smartdatacenter/sdc.py [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+  name:
+    description:
+      - Friendly name for this machine; default is a randomly generated name
+    required: false
+    default: null
+  location:
+    description:
+      - hostname or fqdn for the SDC location you wish to use
+    required: true
+    default: us-east-1
```
Comment (bcoca) [inline cloud/smartdatacenter/sdc.py:36]: if required, default is irrelevant, same for some options below

File: cloud/smartdatacenter/sdc.py [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+DOCUMENTATION = '''
+---
+module: sdc
+short_description: create or delete a virtualmachine or smartmachine in SDC
```
Comment (bcoca) [inline cloud/smartdatacenter/sdc.py:20]: add version_added: 2.0

File: cloud/smartdatacenter/sdc.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+    default: ~/.ssh/id_rsa
+  image:
+    description:
+      - The image UUID to use for this machine
+    required: true
+    default: null
+  package:
+    description:
+      - UUID of the package to use on provisioning
+    required: false
```
Comment (bcoca) [inline cloud/smartdatacenter/sdc.py:60]: when required: False, we do need a default

File: cloud/smartdatacenter/sdc.py [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+  networks:
+    description:
+      - Comma separated list of desired networks ids, obtained from ListNetworks
+    required: false
+  count:
+    description:
+      - An integer value which indicates how many instances that should be created. Only used if count_tag or exact_count are missing (not recommended).
+    required: false
+    default: 1
+    aliases: []
```
Comment (bcoca) [inline cloud/smartdatacenter/sdc.py:70]: omit aliases if empty

File: cloud/smartdatacenter/sdc.py [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+
+    Returns:
+        True if a new machine was created, false otherwise
+    """
+    name = module.params.get('name')
+    package = module.params.get('package')
+    image = module.params.get('image')
+    networks = module.params.get('networks')
+    wait = module.params.get('wait')
+    wait_timeout = int(module.params.get('wait_timeout'))
```
Comment (bcoca) [inline cloud/smartdatacenter/sdc.py:197]: no need to cast if you already requested a type in module definition (as is below)

File: cloud/smartdatacenter/sdc.py [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+        count_needed = exact_count - len(existing_machines)
+    else:
+        count_needed = count
+
+    #Make sure we actually need to create machines, otherwise we can just give info about the existing ones
+    if count_needed > 0:
+        changed = True
+        # Batch create, we'll wait later
+        for x in range(count_needed):
+            sm = sdc.create_machine(name=name, package=package, image=image,
```
Comment (bcoca) [inline cloud/smartdatacenter/sdc.py:220]: no exception handling? I expect sdc to throw exceptions for several reasons, they should be captured and returned through i.e. fail_json(msg=nice message, trackeback=exception)

the same for the other calls to sdc

**Comment by bcoca** [comment]
so mostly minor issues, module looks good in general, if addressed i'll merge soon

**Comment by gregdek** [comment]
@bbhoss if you can address the issues raised by @bcoca we're happy to merge this.

File: cloud/smartdatacenter/sdc.py [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+
+# Provision single machine example
+- local_action:
+    module: sdc
+    account: myaccount
+    key_id: /myaccount/keys/mykey
+    name: my-virtual-machine
+    package: 486bb054-6a97-4ba3-97b7-2413d5f8e849
+    image: 62f148f8-6e84-11e4-82c5-efca60348b9f
+    location: 'us-east-1.api.joyentcloud.com'
```
Comment (dwlf) [inline cloud/smartdatacenter/sdc.py:122]: joyentcloud.com is being deprecated, use joyent.com throughout please.

**Comment by gregdek** [comment]
@bbhoss A friendly reminder: this pull request has been marked as needing your action, and it hasn't been updated in the last month. 

If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we'll keep it open pending your changes.

If we receive no indication from you within a week, we will close this pull request.

**Comment by gregdek** [comment]
Closing due to contributor inaction.

@bbhoss we're happy to see you reopen this PR when you're able to make the necessary revisions.

**Comment by wuservices** [comment]
@bbhoss this looks great! I'm just starting to look at ansible but use Joyent. Did you ever have a chance to continue work on this? Are you still using it?

**Comment by bbhoss** [comment]
I did not. Since I wrote this Terraform has added support for Joyent/Triton/SDC. I would recommend using Terraform to manage instance creation/deletion/etc and Ansible for config management. You should be able to execute Ansible by using local-exec, or just maintaining the hosts file

**Comment by wuservices** [comment]
Thank you. That's great insight and that makes sense. I'll check out Terraform.

**Pull request closed without merging**